feat: Add documentation for `stripCspDirectives` and CSP information (#5194)
cypress-io/cypress-documentation: docs/guides/references/experiments.mdx

| `frame-ancestors`            | No        | Prevents Cypress from loading a test application into an iframe. |
| `require-trusted-types-for`, | No        | Might prevent Cypress from rewriting the DOM.                    |
| `sandbox`                    | Yes       | Can restrict access to script and iframe functionality.          |
| `form-action`                | Yes       | Can prevent Cypress from monitoring form events.                 |
| `navigate-to`                | Yes       | Affects Cypress' ability to navigate to different URLs.          |
| `script-src`                 | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `script-src-elem`            | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `trusted-types`              | No        | Could cause Cypress injections to be marked as untrusted.        |

#### Allow Specific CSP Directives

If you need to test your application with CSP enabled, and need to test against
certain stripped CSP directives, you can set the `experimentalCspAllowList`
option to an array of directive names. This will allow the specified CSP

> directives to be sent to the browser _so long as_ those those directives are

Reviewer: "those" is duplicated here

> directives to be sent to the browser _so long as_ those directives are maked as

Reviewer: It looks like the word pulled into this line should probably be "marked" not "maked"!?

Author: Fixed.

"Allowable". You can find the "Allowable" status of each directive in

> [the table above](#Strip-Minimum-CSP-Directives).

Author: Added

### Experimental CSP Allow List

Cypress will (by default) strip all CSP headers from the response before it is
sent to the browser. This option allows for more granular control over which CSP
directives are stripped from the CSP response headers, allowing you to test your
application with CSP enabled.

#### Strip All CSP Headers

The default value of `experimentalCspAllowList` is `false`. This will remove all
CSP headers from the response before it is sent to the browser. This option
should be used if you do not depend on CSP for any tests in your application.

#### Strip Minimum CSP Directives

If you need to test your application with CSP enabled, you can set the
`experimentalCspAllowList` option to `true`. This will allow all CSP headers to
be sent to the browser _*except*_ those that could prevent Cypress from

> functioning normally. The following CSP directives are automatically stripped

Reviewer 2: I think it would be clear to break the stripped directive table out into the list that is always stripped and then the list that can be configured:

```suggestion
functioning normally. The following CSP directives will always be stripped:

| Stripped Directive           |  Reason                                                           |
| ---------------------------- | ---------------------------------------------------------------- |
| `frame-ancestors`        |    Prevents Cypress from loading a test application into an iframe. |
| `require-trusted-types-for` | Might prevent Cypress from rewriting the DOM.                    |
| `trusted-types`           | Could cause Cypress injections to be marked as untrusted.        |

When `experimentalCspAllowList=true` the following will also be stripped unless specified.

| Stripped Directive           | Allowable | Reason                                                           |
| ---------------------------- | --------- | ---------------------------------------------------------------- |
| `default-src`                | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `sandbox`                    | Yes       | Can restrict access to script and iframe functionality.          |
| `form-action`                | Yes       | Can prevent Cypress from monitoring form events.                 |
| `navigate-to`                | Yes       | Affects Cypress' ability to navigate to different URLs.          |
| `script-src`                 | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `script-src-elem`            | Yes       | Conditionally prevents Cypress from loading scripts and running. |
```

Author: @emilyrohrbough I've made some modificaitons to the overall structure to include this change. Mainly being iterative with an initial table that demonstrates the configuration options and linking to the appropriate sections.

| `script-src`                 | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `script-src-elem`            | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `trusted-types`              | No        | Could cause Cypress injections to be marked as untrusted.        |

#### Allow Specific CSP Directives

If you need to test your application with CSP enabled, and need to test against
certain stripped CSP directives, you can set the `experimentalCspAllowList`
option to an array of directive names. This will allow the specified CSP
directives to be sent to the browser _so long as_ those directives are marked as
"Allowable". You can find the "Allowable" status of each directive in
[the table above](#Strip-Minimum-CSP-Directives).

:::cypress-config-example

```js
{
  experimentalCspAllowList: ['default-src', 'script-src', 'script-src-elem']
}
```

:::

:::caution

Defining `experimentalCspAllowList` _*may*_ cause Cypress to be unable to run
tests against your application. This may be desired if you are testing against
specific CSP directives.

> :::

Reviewer 2: Can you add a note that this experiment is not compatible with the `experimentalSourceRewriting` experiment?

Author: @emilyrohrbough This should still be compatible in certain scenarios. Can we say _may_ not be compatible?

Reviewer 2: @pgoforth @AtofStryker had flagged this as a possible risk / scenario where this configuration opt-in may fail.

- JavaScript is re-written - PR does not handle this scenario
  - **What we do know**: Similar to [SRI tags](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity), nonces/shas need to be provided on inline-scripts CSP into the document before the actual inline script is fetched. If the inline script is rewritten by the proxy, the nonce will work fine, but the sha will break and the script will not execute. This is why when `experimentalModifyObstructiveThirdPartyCode` is enabled, integrity tags are stripped out of HTML documents as it is really difficult to inform the browser of the SHA to expect before the resource is rewritten and SHAed on the server (a bit of the chicken or the egg problem)
  
 Mind adding a blurb stating if there are issues and this may be an known issue

Reviewer 2: @pgoforth I didn't see any code changes in the rewriting that would indicate this wouldn't cause issues with the nonce injection. Did you test any by chance?

Author: @emilyrohrbough Nonce injection is completely independent of SHA hashing. They are handled differently through CSP.

- A SHA hash is an inline validation against the script that gets downloaded and executed. If the version downloaded does not match the SHA hash, the script is not executed.
- A `nonce` value is randomly generated at runtime and does not rely on the file hash at all.

So even if Cypress made inline changes to a script, the `nonce` would still pass CSP validation. We are generating a `nonce` for inline script injections, but are currently not generating a `nonce` for scripts that have been modified by Cypress. `experimentalSourceRewriting` does not trigger a CSP violation by default unless a SHA is used on the script node in question. Therefore, allowing CSP headers is unlikely to cause a the source rewrite code to fail CSP unless there are already specific directive values in place that validate against those scripts. That's why I said it _may_ be an issue, because it's totally dependent on the original CSP headers that Cypress intercepts.

Reviewer 3: the main issue with SHA inline scripts is if the script is rewritten and the sha doesnt match the actual hash. But I think it might be OK to say it may work with `experimentalSourceRewriting` because it is going to depend on the context of what is rewritten, similar to how things work now with the regex rewriter

Author: We can probably specify the scenario (using `'<hash-algorithm>-<base64-value>'` inside a CSP directive) as that's the only place it could trip against `experimentalSourceRewriting` or `experimentalModifyObstructiveThirdPartyCode`/`modifyObstructiveCode`.

I also genuinely feel like that is a problem that can be solved given our ability to modify individual directive values. We would just need to match the original SHA in the CSP directive with the one we replaced, and swap the value for the replacement. Outside of the scope of this PR, but definitely doable in the near future.

| `default-src`      | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `sandbox`          | Yes       | Can restrict access to script and iframe functionality.          |
| `form-action`      | Yes       | Can prevent Cypress from monitoring form events.                 |
| `navigate-to`      | Yes       | Affects Cypress' ability to navigate to different URLs.          |
| `script-src`       | Yes       | Conditionally prevents Cypress from loading scripts and running. |
| `script-src-elem`  | Yes       | Conditionally prevents Cypress from loading scripts and running. |

#### Allow Specific CSP Directives

You can also set the `experimentalCspAllowList` option to an array of directive
names marked as "Allowable". This will allow the specified CSP directives to be
sent to the browser.

The following configuration would allow the `default-src`, `script-src`, and
`script-src-elem` directives to be sent to the browser:

:::cypress-config-example

```js
{
  experimentalCspAllowList: ['default-src', 'script-src', 'script-src-elem']
}
```

:::

:::caution

> Defining `experimentalCspAllowList` _*may*_ cause Cypress to be unable to run
> tests against your application. This may be desired if you are testing against
> specific CSP directives.

Reviewer 2: I don't think this would ever be desirable if someone is trying to automate tests with Cypress 😅 Maybe:
```suggestion
Defining `experimentalCspAllowList` _*may*_ cause Cypress to be unable to run
tests against your application. If you experience issues, reduce the directives specified in your allow list to identify which direction is causing issues.
```

Author: So yes, I understand the concern. In the previous itteration, any directive was allowed to be provided, which I could find valid reasons to test against. In the current state, it's doubtful that the directives we are permitting will cause Cypress to be unable to run. The ones that could badly conflict are the ones that are never allowed.

I do like the wording you've suggested, because it gives a path forward for resolution.

| `trusted-types`             | No        | Could cause Cypress injections to be marked as untrusted.        |

When `experimentalCspAllowList=true` the following directives will also be
stripped in addition to the ones listed above:

| Stripped Directive | Allowable | Reason                                                                        |
| ------------------ | --------- | ----------------------------------------------------------------------------- |
| `child-src`        | Yes       | Could prevent iframes from loading in combination with other Cypress options. |
| `default-src`      | Yes       | Conditionally prevents Cypress from loading scripts and running.              |
| `frame-src`        | Yes       | Could prevent iframes from loading in combination with other Cypress options. |
| `form-action`      | Yes       | Can prevent Cypress from monitoring form events.                              |
| `script-src`       | Yes       | Conditionally prevents Cypress from loading scripts and running.              |
| `script-src-elem`  | Yes       | Conditionally prevents Cypress from loading scripts and running.              |

#### Allow Specific CSP Directives

> You can also set the `experimentalCspAllowList` option to an array of directive

Reviewer 2: ```suggestion
Set the `experimentalCspAllowList` option to an array of directive
```
